Sources/Basic/TemporaryFile.swift

```diff
@@ -0,0 +1,172 @@
+    ///
+    /// - Throws: TempFileError
+    public init(dir: String? = nil, prefix: String = "", suffix: String = "") throws {
+        self.suffix = suffix
+        self.prefix = prefix
+        // Determine in which directory to create the temporary file.
+        self.dir = TemporaryFile.determineTempDirectory(dir)
+        // Construct path to the temporary file.
+        let path = self.dir + prefix + "TemporaryFile.XXXXXX" + suffix
+
+        // Convert path to a C style string terminating with null char to be an valid input
+        // to mkstemps method. The XXXXXX in this string will be replaced by a random string
+        // which will be the actual path to the temporary file.
+        var template = ByteString(encodingAsUTF8: path).contents.map { Int8($0) } + [Int8(0)]
+
+        fd = libc.mkstemps(&template, Int32(suffix.characters.count))
+        // If mkstemps failed then throw error.
+        if fd == -1 { throw TempFileError(errno: errno) }
+
+        self.path = String(cString: template)
+        fileHandle = NSFileHandle(fileDescriptor: fd)
+    }
+
+    /// Convinence method for writing to the temporary file using OutputByteStream.
+    /// Note: This will write using the fileHandle and hence move the handle's filepointer.
+    ///
+    /// - Parameters:
+    ///     - write: Closure which provides the OutputByteStream object to write to.
+    ///
+    /// - Throws: SystemError
+    public func write(write: (OutputByteStream) -> Void) throws {
+        write(stream)
+        return try fputs(stream.bytes.contents, fileHandle)
+    }
+
+    /// Remove the temporary file before deallocating.
+    deinit { unlink(path) }
+
+    /// Create a default TemporaryFile object for duration of the closure passed to it.
+    ///
+    /// - Parameters:
+    ///     - body: Closure providing a default instance of TemporaryFile.
+    ///
+    /// - Throws: TempFileError
+    public static func withFile<T>(body: @noescape (TemporaryFile) throws -> T) throws -> T {
+        return try body(try TemporaryFile())
+    }
+
+    /// Determines the directory in which the temporary file should be created. Also makes
+    /// sure the returning path has a trailing forward slash.
+    ///
+    /// - Parameters:
+    ///     - dir: If present this will be the temporary directory.
+    ///
+    /// - Returns: Path to directory in which temporary file should be created.
+    private static func determineTempDirectory(_ dir: String? = nil) -> String {
+        // FIXME: Add other platform specific locations.
+        var tmpDir = dir ?? getenv("TMPDIR") ?? getenv("TEMP") ?? getenv("TMP") ?? "/tmp/"
+        if !tmpDir.hasSuffix("/") { tmpDir += "/" }
+        precondition(tmpDir.isDirectory)
+        return tmpDir
+    }
+
+    /// Read the temporary file contents as a UTF8 encoded string if possible.
+    /// This will use a new file handle to read the file and not touch the file handle
+    /// exposed by this class.
```

Comment:
Why is that part of the API? This is actually a security hole -- part of the reason that the system `mk*temp` functions are designed as they are is explicitly to avoid the potential for an attacker to gain the ability to inject contents into the temporary file -- using path based operations after receiving the path opens up this hole again.